### PR TITLE
Implement Particle.cancel()

### PR DESCRIPTION
**Pull request by sergeuz**
### Problem

This PR implements a method called `Particle.cancel()` which closes the cloud connection synchronously, without waiting for any unacknowledged messages.

### Steps to Test

_TBD_

### References

- [ch42232]

### CODE REVIEW

File: wiring/src/spark_wiring_cloud.cpp [hunk 1]
```diff
@@ -76,3 +76,9 @@ Future<bool> CloudClass::publish_event(const char *eventName, const char *eventD
 int CloudClass::publishVitals(system_tick_t period_s_) {
     return spark_publish_vitals(period_s_, nullptr);
 }
+
+void CloudClass::cancel() {
```
Comment (m-mcgowan) [inline wiring/src/spark_wiring_cloud.cpp]: I would hold off adding a new public API for a bugfix release and instead give the customer this code block as an example of how to synchronously cancel the cloud connection. Let's also document the expected behavior, specifically that the ongoing connection isn't actively interrupted, but will continue to the next logical step. E.g. if the handshake is ongoing then the handshake will continue to completion or failure, but will not be prematurely cancelled.

File: system/inc/system_cloud.h [hunk 2]
```diff
@@ -240,6 +240,23 @@ void spark_cloud_flag_disconnect(void);    // should be set connected since it m
  */
 bool spark_cloud_flag_auto_connect(void);
 
+/**
+ * Option flags changing the behavior of `spark_cloud_disconnect()`.
+ */
+typedef enum cloud_disconnect_flag {
+    CLOUD_DISCONNECT_GRACEFULLY = 0x01, ///< Disconnect gracefully.
+    CLOUD_DISCONNECT_DONT_CLOSE = 0x02 ///< Do not close the socket.
```
Comment (m-mcgowan) [inline system/inc/system_cloud.h]: What does it mean to disconnect without closing the socket?

File: wiring/src/spark_wiring_cloud.cpp (hunk 1 above)
Comment (sergeuz) [reply wiring/src/spark_wiring_cloud.cpp]: I renamed this method to `Particle._cancel()` as we discussed.

File: system/inc/system_cloud.h (hunk 2 above)
Comment (sergeuz) [reply system/inc/system_cloud.h]: It doesn't make much sense from the user's perspective, but we use it internally as an optimization. These flags are also defined in [#1899](https://github.com/particle-iot/device-os/pull/1899/files#diff-b43faa0ec0ffb23fa014851243cadf77R87), so having the same set of flags in this PR will simplify merging.

**Pull request closed without merging**